flo-runner/src/main/java/com/spotify/flo/context/ForkingEvalContext.java

```diff
@@ -0,0 +1,385 @@
+        }
+      } finally {
+        executor.shutdown();
+        tryDeleteDir(tempdir);
+      }
+    };
+  }
+
+  private void tryDeleteDir(Path path) {
+    try {
+      deleteDir(path);
+    } catch (IOException e) {
+      LOG.warn("Failed to delete directory: {}", path, e);
+    }
+  }
+
+  private void deleteDir(Path path) throws IOException {
+    Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
+      @Override
+      public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+        Files.delete(file);
+        return FileVisitResult.CONTINUE;
+      }
+
+      @Override
+      public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+        Files.delete(dir);
+        return FileVisitResult.CONTINUE;
+      }
+    });
+  }
+
+  private void copyLines(InputStream in, PrintStream out) {
+    final BufferedReader reader = new BufferedReader(new InputStreamReader(in));
+    try {
+      String line;
+      while ((line = reader.readLine()) != null) {
+        out.println(line);
+      }
+    } catch (IOException e) {
+      log.error("Caught exception during stream copy", e);
+    }
+  }
+
+  @SuppressWarnings("finally")
+  private static class Trampoline {
+
+    private static final boolean DEBUG = Boolean.parseBoolean(System.getenv("FLO_DEBUG_FORKING"));
+
+    private static final String NAME = ManagementFactory.getRuntimeMXBean().getName();
+
+    private static class Watchdog
+        extends Thread {
+
+      Watchdog() {
+        setDaemon(false);
+      }
+
+      @Override
+      public void run() {
+        // Wait for parent to exit.
+        try {
+          while (true) {
+            int c = System.in.read();
+            if (c == -1) {
+              break;
+            }
+          }
+        } catch (IOException e) {
+          errPrefix();
+          e.printStackTrace(System.err);
+          System.err.flush();
+        }
+        System.err.println();
+        err("child process exiting");
+        // Exit with non-zero status code to skip shutdown hooks
+        System.exit(-1);
+      }
+    }
+
+    public static void main(String... args) {
+      err("child process started");
+      Watchdog watchdog = new Watchdog();
+      watchdog.start();
+
+      final TaskId taskId;
+      try {
+        taskId = TaskId.parse(System.getenv("FLO_TASK_ID"));
+      } catch (IllegalArgumentException e) {
+        try {
+          e.printStackTrace();
+          System.err.flush();
+        } finally {
+          System.exit(1);
+        }
+        return;
+      }
+      err("read FLO_TASK_ID = " + taskId);
+
+      if (args.length != 3) {
+        err("args.length != 3");
+        System.exit(1);
+        return;
+      }
+      final Path closureFile = Paths.get(args[0]);
+      final Path resultFile = Paths.get(args[1]);
+      final Path errorFile = Paths.get(args[2]);
+
+      Context.current().withValue(Tracing.TASK_ID, taskId).run(() ->
```

Comment:
How does this solve other thread than main outputting logs?

Reply:
Or the encoder will check both environment and context?

Reply:
I guess I don't want to break the feature that the task-id is available in the task process fn, partly because it's already on master (we merged that PR and there's tests for it) and it might still come in handy.